src/krb5_plugin/passkey/passkey_utils.c

```diff
@@ -0,0 +1,556 @@
+        if (data == NULL) {
+            goto done;
+        }
+        break;
+    case SSS_PASSKEY_PHASE_REPLY:
+        data = sss_passkey_reply_from_json_object(jdata);
+        if (data == NULL) {
+            goto done;
+        }
+        break;
+    default:
+        goto done;
+    }
+
+    message = sss_passkey_message_init(phase, state, data);
+    if (message == NULL && phase == SSS_PASSKEY_PHASE_CHALLENGE) {
+        sss_passkey_challenge_free(data);
+    } else if (message == NULL && phase == SSS_PASSKEY_PHASE_REPLY) {
+        sss_passkey_reply_free(data);
+    }
+
+done:
+    json_decref(jroot);
+    return message;
+}
+
+static char *
+sss_passkey_message_to_json(const struct sss_passkey_message *message)
+{
+    json_t *jroot;
+    json_t *jdata;
+    char *str;
+
+    if (message == NULL) {
+        return NULL;
+    }
+
+    switch (message->phase) {
+    case SSS_PASSKEY_PHASE_INIT:
+        if (message->state != NULL || message->data.ptr != NULL) {
+            return NULL;
+        }
+        jdata = NULL;
+        break;
+    case SSS_PASSKEY_PHASE_CHALLENGE:
+        if (message->state == NULL || message->data.challenge == NULL) {
+            return NULL;
+        }
+
+        jdata = sss_passkey_challenge_to_json_object(message->data.challenge);
+        if (jdata == NULL) {
+            return NULL;
+        }
+        break;
+    case SSS_PASSKEY_PHASE_REPLY:
+        if (message->state == NULL || message->data.reply == NULL) {
+            return NULL;
+        }
+
+        jdata = sss_passkey_reply_to_json_object(message->data.reply);
+        if (jdata == NULL) {
+            return NULL;
+        }
+        break;
+    default:
+        return NULL;
+    }
+
+    jroot = json_pack("{s:i, s:s?, s:o?}",
+                      "phase", message->phase,
+                      "state", message->state,
+                      "data", jdata);
+    if (jroot == NULL) {
+        json_decref(jdata);
+        return NULL;
+    }
+
+    str = json_dumps(jroot, JSON_COMPACT);
+    json_decref(jroot);
+
+    return str;
+}
+
+char *
+sss_passkey_message_encode(const struct sss_passkey_message *data)
+{
+    return sss_radius_message_encode(SSSD_PASSKEY_PREFIX,
+        (sss_radius_message_encode_fn)sss_passkey_message_to_json, data);
+}
+
+struct sss_passkey_message *
+sss_passkey_message_decode(const char *str)
+{
+    return sss_radius_message_decode(SSSD_PASSKEY_PREFIX,
+        (sss_radius_message_decode_fn)sss_passkey_message_from_json, str);
+}
+
+krb5_pa_data *
+sss_passkey_message_encode_padata(const struct sss_passkey_message *data)
+{
+    return sss_radius_encode_padata(SSSD_PASSKEY_PADATA,
+        (sss_radius_message_encode_fn)sss_passkey_message_encode, data);
+}
+
+struct sss_passkey_message *
+sss_passkey_message_decode_padata(krb5_pa_data *padata)
+{
+    return sss_radius_decode_padata(
+        (sss_radius_message_decode_fn)sss_passkey_message_from_json, padata);
```

Comment:
I think `sss_passkey_message_decode` would be better here so that the `passkey ` prefix, which you mentioned in the description for the message, is removed.